File: tests/fixtures/stamp_defaults.py
# Fixture for the LANGUAGE half of the stamp-only applicable set.
#
# Python declares no address-of operator, so EscapeEnricher returns before it
# reads anything and no Python function has ever been examined for escaping
# locals. Every function here must therefore answer NEITHER value for
# has_escape — not 'true', and not the default either.
#
# The other three enrichers DO run on Python: it declares a comment kind and a
# call expression, and the shadow enricher gates on the node kind and nothing
# else. So the very same rows answer 'true' or 'false' for has_todo,
# is_recursive and has_shadow. One file, one language, four fields, and the
# fields disagree — which is the whole claim.


def plain_function(x):
    return x + 1


def function_with_todo(x):
    # TODO: this one is stamped, so it is not a default
    return x + 2


def recursive_function(x):
    if x <= 0:
        return 0
    return recursive_function(x - 1)


def shadowing_function(x):
    total = 0

    def inner():
        total = 1
        return total

    return inner() + total

File: tests/fixtures/stamp_defaults.cpp
// Fixture for the stamp-only boolean defaults.
//
// has_todo, has_escape, has_shadow and is_recursive are written only when they
// hold. The other value is not stored anywhere, so it is answered from the
// declaration in the field-tier table: the rows the enricher examined, minus
// the rows it wrote.
//
// Four functions and one struct, and the struct is named to sort LAST. It is
// the load-bearing part twice over: a row of a kind no function enricher
// examines must answer NEITHER value, and — because "" sorts before "false" —
// its position under ORDER BY is the one place the resolver's effect on sorting
// is visible at all. With a two-valued field the unstamped rows sort where the
// valueless ones used to, so a fixture of functions alone could not tell the
// two readings apart.

struct zNotAFunction {
    int member;
};

int plainFunction(int x)
{
    return x + 1;
}

int functionWithTodo(int x)
{
    // TODO: this one is stamped, so it is not a default
    return x + 2;
}

int recursiveFunction(int x)
{
    if (x <= 0) {
        return 0;
    }
    return recursiveFunction(x - 1);
}

int shadowingFunction(int x)
{
    int value = x;
    {
        int value = 2;
        x += value;
    }
    return x + value;
}
